[sw, tock] Extend downstream tock development infrostructure

silvestrst opened the pull request:
The upstream board/chip split has been preserved. The ibex crate has
    been renamed to earlgrey, and lowrisc to opentitan_common. In future
    OpenTitan could have more than one chip implementation.
    opentitan_common crate is a collection of drivers that can be shared
    between different OpenTitan chips. The instantiation of these drivers
    is done inside a specific chip crate (earlgrey).
    
    The upstream/external tock dependencies have been split out in their
    own "tock" crate. This allows us to avoid having local/remote
    Cargo.toml pair every time we need to use a remote tock dependency
    inside a local crate.
    
    The second commit adds an option to use device specific configuration.

    NOTE:
    Because our current boards/opentitan downstream implementation is
    behind the upstream, the revision of external components is specified
    directly. When the downstream implementation is refactored to use
    up-to-date tock API, this can be dropped, and from there we can rely
    on the Cargo.lock alone.

NOTE-2: The approach is not to change any existing functionality copied from the upstream `ibex` and `lowrisc` creates. The drivers will change drastically with refactoring to use DIFs, and it would be wasteful to do any cleanup now.

Initial discussion and the draft PR #2087 .
Next PR in series is #2119 .

**Please have a look at the corresponding issue (#2139 ) before reviewing this change. It has complete outline of the problem.**

jon-flatley commented:
I don't think we want to remove the `Cargo_remote.lock` file if we're going to stick with the symlink approach. Right now when you do a build it generates a dead symlink in `sw/device/tock/tock/Cargo.lock` that points to `Cargo_remote.lock`, which doesn't exist and doesn't get generated. Including a `Cargo_remote.lock` could also replace the use of `rev=` in the `Cargo_remote.toml` dependencies.

silvestrst commented:
> I don't think we want to remove the Cargo_remote.lock file if we're going to stick with the symlink approach.

`tock` crate is built as a library, cargo doesn't produce lock files for libraries.

> Including a Cargo_remote.lock could also replace the use of rev= in the Cargo_remote.toml dependencies.

Well, the only lock file is in `boards/opentitan`, and that is under the source control.

`rev` is not a permanent solution, it is just an easy way to force a specific version of the tock for now. If I remove `rev` cargo will pull the upstream master, which will break our `board`, possibly `chip` as well - due to API change in upstream kernel interfaces.

We should sync it up, but it should be a follow-up (requires refactoring existing board implementation).

silvestrst commented:
Does anyone know whether there is any advantage of using `#![crate_type = "rlib"]` in `lib.rs`, instead of specifying it in `Cargo.toml` through `[lib]`?

silvestrst commented:
* Updated crate author names.
* Removed unnecessary entries from `.gitignore`.
* Removed unnecessary `.lock` symlink magic from `tock/meson.build `.

mcy commented:
@silvestrst: the values in `#[crate_*]` attributes are usually passed to rustc via command line flags by cargo. You only need them when you're doing really silly stuff to build your Rust (usually, if you're not using Cargo).

You probably don't need them.

lenary commented on util/build_tock.sh:
```diff
@@ -24,4 +24,5 @@ cargo +"$(cat ${TOOLCHAIN_FILE})" build \
 	--target "${TARGET}" \
 	--manifest-path "${MANIFEST_PATH}" \
 	--target-dir "${TARGET_DIR}" \
+  --features="${TOCK_FEATURES}" \
```
Nit: indentation, presumably this file uses tabs, though I'm not sure why.

silvestrst replied:
Done.

silvestrst commented on sw/device/tock/chips/earlgrey/README.md:
```diff
@@ -0,0 +1,5 @@
+Earlgrey chip
+============================
+
+OpenTitan chip implementation.
```
Good point, will do.

silvestrst replied:
Done.

silvestrst commented on sw/device/tock/chips/earlgrey/src/device_config.rs:
```diff
@@ -0,0 +1,22 @@
+}
+
+// This is also a default case, when no device is specified.
+#[cfg(any(feature = "device_fpga", not(feature = "device_specified")))]
+pub const DEVICE_CONFIG: DeviceConfig = DeviceConfig {
+    name: &"device_fpga",
+    chip_freq: 50_000_000,
+    uart_baudrate: 230400,
+};
+
+#[cfg(feature = "device_verilator")]
+pub const DEVICE_CONFIG: DeviceConfig = DeviceConfig {
+    name: &"device_verilator",
+    chip_freq: 500_000,
+    uart_baudrate: 9600,
+};
```
Hey @alphan , thank you for your review.

> But, I think it would be better if you could add a definition that fails to compile for the the case when no device is specified (instead of silently using device_fpga).

Can do, the reason why I have done it this way - is to do with upstream CI, which builds every crate separately. This means that we can't put `default` feature in the crate (default feature is always built, and will conflict if we pass another option), so essentially I have implemented the default case in code instead. For downstream, we can make it fail - but then effectively the default case moves to the point of `build_tock.sh` or `meason.build` level. I think these improvements should be done at the later stage/follow-up (the second stage #2139 ).

> Also, can you match these to the names used in sw/device/lib/arch/meson.build, i.e. sim_verilator and fpga_nexysvideo.

Sure, will do.

> I find it a bit counterintuitive to name the chip-specific directory earlgrey and the shared-code directory opentitan_common.

I can see that, but in my opinion it is not an unreasonable split, I see OpenTitan as a project, and earlgrey a specific chip implementation of this project. It is true that we don't need this split at the moment, as we only have one chip.

> If there are no other crates that depend on opentitan_common, maybe we should hold off on this split until we need it.

This is essentially what the upstream `ibex` and `lowrisc` crates are, I am happy to rename `opentitan_common` to something better. Initially I had it names `peripherals`, but then that doesn't work if we eventually decide to upstream it.

The only alternative I see, is to stick `opentitan_common` under `earlgrey`, either as a create or just put peripheral `.rs` files there. As they are more or less copies of the `ibex` and `lowrisc` crates in upstream, it would be easier to the merge (if necessary) in the second stage (#2139 ).

> Is there a reason why the code for plic and timer is not moved to opentitan_common
This is how these peripherals were implemented in the upstream, and further peripheral refactoring is subject to the second stage of the Tock downstream development (#2139 )

The idea is to take smaller steps at the time, there are plenty of improvements and tidying up that can be done in follow-ups.

silvestrst commented:
* Addressed @lenary comments.
* Added copyright to new files, including Cargo.toml.
* Renamed `device_config.rs` to `chip_config.rs` (it could be a nice place to instantiate the `opentitan_common` peripherals in here. Eventually ...).
* Addressed one of @alphan comments:
> Also, can you match these to the names used in sw/device/lib/arch/meson.build, i.e. sim_verilator and fpga_nexysvideo.

* Addressed @mcy comment (https://github.com/lowRISC/opentitan/pull/2129#pullrequestreview-405878515).

asb commented:
Two related, minor requests:
* Could you please move sw/device/tock/boards/opentitan/COPYRIGHT to sw/device/tock
* Could you please add a README to sw/device/tock to explain what lives in this sub-tree in a similar way you've done so in other subdirs and perhaps explain that files copied in from Tock don't have the standard lowRISC headers, and refer people to COPYRIGHT for licensing information.

asb commented on sw/device/tock/README:
```diff
@@ -0,0 +1,43 @@
+Tock OS
```
Please move to README.md so this renders nicely in the GitHub UI.

We tend to use the more common `# Heading`, `## Subheading` syntax for headings in Markdown too

silvestrst replied:
done.

lenary commented on sw/device/tock/README:
```diff
@@ -0,0 +1,43 @@
+
+Copyright
+=========
+Any files taken directly from the upstream must keep their original copyright.
+
+Any new files must have lowRISC copyright, please see `COPYRIGHT`.
```
This section is a bit strange. Can you rewrite it as:

> The source files in this directory come from two sources, with slightly different copyright licences.
>
> Some source files are copied from the upstream tock repository. The notice in `COPYRIGHT` in this directory applies to all source files with explicit Tock OS Developers copyright headers, or no copyright headers at all.
>
> Where lowRISC has added additional files, these are marked with the lowRISC Copyright header, and the notice in `LICENCE` in the root of this repository applies.

silvestrst replied:
Done.

asb commented on sw/device/tock/README:
```diff
@@ -0,0 +1,43 @@
+
+Copyright
+=========
+Any files taken directly from the upstream must keep their original copyright.
```
I suggest replacing the two sentencies in this section with "Some files in this subdirectory are imported directly from upstream Tock, so do not include our standard license header. See COPYRIGHT for the original upstream copyright notice."

silvestrst replied:
Already applied Sams suggestion (probably whilst you were writing this comment).

asb commented on sw/device/tock/chips/earlgrey/README.md:
```diff
@@ -0,0 +1,7 @@
+Earlgrey chip
+============================
```
Slightly prefer `# Heading`

silvestrst replied:
done.

asb commented on sw/device/tock/chips/opentitan_common/README.md:
```diff
@@ -0,0 +1,6 @@
+OpenTitan common peripherals
+============================
```
Slightly prefer `# Heading`

silvestrst replied:
done.

asb commented on sw/device/tock/README.md:
```diff
@@ -0,0 +1,46 @@
+# Tock OS
+
```
Apologies, I tired to request this before but looking back it may have got lost. Could you please add a single sentence description either of the first or second sentence that explains what is contained in this subdirectory, in a way that makes sense for someone unfamiliar with Tock.

A reasonable assumption when navigating to sw/device/tock is that you might find a complete copy of the Tock OS. It would be good to quickly dispel that misconception.

silvestrst replied:
Didn't quite manage a single sentence, but added to the description.

silvestrst replied:
Done.

gkelly commented on sw/device/tock/chips/earlgrey/src/lib.rs:
```diff
@@ -0,0 +1,15 @@
+//! Drivers and chip support for the Ibex soft core.
```
Outdated comment.

silvestrst replied:
Ibex -> Earlgrey, done.

gkelly commented on sw/device/tock/chips/opentitan_common/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+
+[package]
+name = "opentitan_common"
```
I think these should just be `opentitan`.

silvestrst replied:
Unfortunately the name of board level `opentitan` package clashes with this:
`error: package collision in the lockfile: packages opentitan v0.1.0 (/home/svt/development/repos/opentitan/sw/device/tock/boards/opentitan) and opentitan v0.1.0 (/home/svt/development/repos/opentitan/sw/device/tock/chips/opentitan) are different, but only one can be written to lockfile unambiguously`

Meanwhile I will just change the name of the board level name to "opentitan_board", nothing depends on a top level package at the moment, and we can always sort it out in a follow-up.

gkelly replied:
I'd say let's leave this as is for now and we can change it later if it grates on us too much. 😉

silvestrst commented on sw/device/tock/boards/opentitan/src/counter_demo.rs:
```diff
@@ -4,6 +4,8 @@
 
 //! A simple alarm implementation to display a binary counter over GPIO.
 
+use tock::kernel as kernel;
```
I think you might been looking at the slightly outdated version of the change, in the latest PR it has already been fixed.

silvestrst commented:
Have addressed all of the outstanding comments, apart from renaming "opentitan_common" to "opentitan" (see the explanation above).

silvestrst commented on sw/device/tock/chips/opentitan_common/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+edition = "2018"
+
+[dependencies]
+tock = { path = "../../tock" }
```
Nit: add newline

silvestrst replied:
Done.

asb commented on sw/device/tock/README.md:
```diff
@@ -0,0 +1,52 @@
+This is not a complete Tock OS implementation. OpenTitan relevant contents of
+`board` and `chip` directories are implemented and maintained downstream, the
+other Tock OS dependencies are pulled from the upstream or local Tock OS repo.
+Please see [Rationale](# Rationale) and [Directory layout](# Directory layout)
```
These links don't render. It's a short doc, so I'd just say "See below for more details." or similar.

silvestrst replied:
Done.

asb commented on sw/device/tock/README.md:
```diff
@@ -0,0 +1,52 @@
+
+Tock OS is a secure embedded operating system for microcontrollers.
+
+This is not a complete Tock OS implementation. OpenTitan relevant contents of
```
This could be misread as being a statement about the level of progress in porting Tock. How about keeping it shorter with something like "This directory contains OpenTitan board and chip configurations for Tock that are maintained downstream."

silvestrst replied:
now it reads as follows:
```
This directory contains OpenTitan board and chip implementation for Tock that
are maintained downstream. Other OS components are pulled directly from
upstream or a local Tock OS repository. See below for more details.
```
